### DISCUSSION TITLE
[WIP] Shinkansen

### DISCUSSION THREAD

**Discussion by Milo123459** [railwayapp/nixpacks#340] [RFCs]
Shinkansen is a system that would allow nixpacks to offer ancient and deprecated versions of packages previously published to nix. Shinkansen would work similarly to something like [this](https://lazamar.co.uk/nix-versions/?channel=nixpkgs-unstable&package=nodejs). It will give back a nixpkg archive and a package name to use.

Shinkansen would be a monorepo(tbd) containing of two crates: `shinkansen_api` and `shinkansen`. `shinkansen_api` does the fetching of nixpkgs and resolves them to hashes and versions, and `shinkansen` is the actual webserver that has aggressive caching.

### Functions

- get all available package versions (for debugging, wouldn't actually be used in nixpacks)
- get a nix archive and package name to use based on a package version

**Comment by coffee-cup** [comment]
> for debugging, wouldn't actually be used in nixpacks

Not necessarily. I can see a few use cases for it in the future. Specifically for showing available versions if the version the user requested is not available.

Also you have the names backwards. `shinkansen` would be the library/package that interacts with nixpkgs. The api would be a way to interact with the library.

Initially, just the library can be worked out without worrying about the API until the functionality is complete.

Lastly, I am not sure I like the name 🙈 It doesn't really relate to Nixpacks or Nix at all.

**Reply by Milo123459** [reply]
I like the idea of showing available versions if none can be matched. The naming stuff is fine, that can always be changed.

Regarding the name, that's fine! We just need to find something we can all agree on haha
